src/strands/tools/registry.py

```diff
@@ -24,244 +24,318 @@
+        """
+        if tool_name is None:
+            raise ValueError("Tool name cannot be None")
+
+        # Check if the tool exists before attempting to delete it
+        if tool_name not in self.agent_tools:
+            raise ValueError(f"Tool '{tool_name}' not found in registry")
+
+        # Make a backup of the tool in case we need to restore it
+        original_tool = self.agent_tools.get(tool_name)
+        original_hot_reload = self._hot_reload_tools.get(tool_name)
+
+        # Find and remove the tool from the tool config
+        original_tool_config_entry = None
+        for idx, tool_entry in enumerate(self._tool_config.get("tools", [])):
+            if tool_entry.get("toolSpec", {}).get("name") == tool_name:
+                original_tool_config_entry = (idx, tool_entry)
+                break
 
-            if matching_tools:
-                raise ValueError(
-                    f"Tool name '{tool.tool_name}' already exists as '{matching_tools[0]}'."
-                    " Cannot add a duplicate tool which differs by a '-' or '_'"
-                )
+        try:
+            # Delete the tool from the main registry
+            del self.agent_tools[tool_name]
+            logger.debug("tool_name=<%s> | deleted from main registry", tool_name)
+
+            # Delete from hot reload tools if present
+            if tool_name in self._hot_reload_tools:
+                del self._hot_reload_tools[tool_name]
+                logger.debug("tool_name=<%s> | deleted from dynamic tools", tool_name)
+
+            # Remove from tool config if present
+            if original_tool_config_entry:
+                idx, _ = original_tool_config_entry
+                self._tool_config["tools"].pop(idx)
+                logger.debug("tool_name=<%s> | removed from tool config", tool_name)
+
+            return tool_name
+        except Exception as e:
+            # Restore the tool if deletion fails
+            if original_tool:
+                self.agent_tools[tool_name] = original_tool
+            if original_hot_reload:
+                self._hot_reload_tools[tool_name] = original_hot_reload
+
+            # Restore tool config entry if it was removed
+            if original_tool_config_entry:
+                idx, entry = original_tool_config_entry
+                if idx < len(self._tool_config["tools"]):
+                    self._tool_config["tools"].insert(idx, entry)
+                else:
+                    self._tool_config["tools"].append(entry)
 
-        # Register in main registry
-        self.registry[tool.tool_name] = tool
+            logger.warning("tool_name=<%s> | failed to delete tool | %s", tool_name, e)
+            raise ValueError(f"Failed to delete tool '{tool_name}': {str(e)}") from e
 
-        # Register in dynamic tools if applicable
-        if tool.is_dynamic:
-            self.dynamic_tools[tool.tool_name] = tool
+    def list_tools(self) -> Dict[str, ToolSpec]:
```

Comment:
It's a little weird that a `list` method is returning a dictionary.  Would this be better named as `get_tool_specs` or something?